src/python/deepgnn/graph_engine/snark/distributed.py

```diff
@@ -43,7 +46,38 @@ def __init__(
         stream: bool = False,
     ):
         """Init snark server."""
-        with open(os.path.join(data_path, "meta.txt"), "r") as meta:
+        path = data_path
+
+        if (
+            path.startswith("hdfs://")
+            or path.startswith("adl://")
+            or path.startswith("file:///")
+        ):
+            _set_hadoop_classpath(config_path)
+
+            class _ErrCallback:  # Copied from client.py
+                def __init__(self, method: str):
+                    self.method = method
+
+                # We have to use mypy ignore, to reuse this callable object across
+                # all C function call because they have different signatures.
+                def __call__(self, result, func, arguments):
+                    if result != 0:
+                        raise Exception(f"Failed to {self.method}")
+
+            lib = _get_c_lib()
+            lib.HDFSMoveMeta.errcheck = _ErrCallback("hdfs move meta")  # type: ignore
+
+            hdfs_path = path
+            temp_dir = tempfile.TemporaryDirectory()
+            path = temp_dir.name
+            lib.HDFSMoveMeta(
+                c_char_p(bytes(_get_meta_path(hdfs_path), "utf-8")),
+                c_char_p(bytes(_get_meta_path(path), "utf-8")),
+                c_char_p(bytes(config_path, "utf-8")),
+            )
```

Comment:
Since this is duplicated code with meta.py, can we make it a utility function in meta.py instead.

Reply:
Also pull main for tf_examples fix.

Reply:
Done, thanks!